### PR TITLE
Overlay implementation

### LINKED ISSUE #1387: igx-Overlay
Implement igxOverlay service

### Research 
- [x] Research document
- [x] Technical specification
- [x] Unit tests
- [x] POC

### Development
- [x] POC polish
- [x] Implementation
   - [x] Modal option
   - [x] Directive rename igx-Toggle to igxOverlayDirective 
   - [x] Auto position strategy
   - [x] Global strategy
   - [x] Connected position strategy
- [ ] Styling & UI implementation
### Tests

- [ ] Unit tests
- [ ] User scenarios tests 
### Documentation
- [x] ReadMe
- [ ] DocFX
- [ ] Samples creation
- [ ] Changelog update
- [x] API Code Snippets

### Links
- [Development branch](https://github.com/IgniteUI/igniteui-angular/tree/overlay-implementation)
- [Research document](https://infragisticsinc297.sharepoint.com/:w:/r/Groups/ProdDev/_layouts/15/Doc.aspx?sourcedoc=%7B83d6d6fd-bb56-439d-8699-f17f4829df46%7D&action=default)
- [Technical specification](https://github.com/IgniteUI/igniteui-angular/wiki/Overlay-Specification)

### PR DESCRIPTION
**Pull request by wnvko**
Closes #1387

### CODE REVIEW

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+            componentRef: dc,
+            size: element.getBoundingClientRect(),
+            settings: overlaySettings
+        });
+        return element;
+    }
+
+    private getPositionStrategy(positionStrategy?: IPositionStrategy): IPositionStrategy {
+        positionStrategy = positionStrategy ? positionStrategy : new GlobalPositionStrategy();
+        if (positionStrategy._settings && positionStrategy._settings.element) {
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: Make this a generic utility static method that takes the settings and returns a point form the target (either directly, or doing these calculations based on element)

File: projects/igniteui-angular/src/lib/services/overlay/scroll/AbsoluteScrollStrategy.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        componentRef: ComponentRef<{}>,
+        size: { width: number, height: number },
+        settings: OverlaySettings
+    };
+
+    constructor(scrollContainer?: HTMLElement) {
+        this._scrollContainer = scrollContainer;
+    }
+
+    initialize(document: Document, overlayService: IgxOverlayService, id: string) {
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/services/overlay/scroll/AbsoluteScrollStrategy.ts]: How about we simplify the `IScrollStrategy` interface (without `initialize`, you can pass in the service + id in the constructor if needed).
Plus add a `[re]position(id: string)` method to the service to remove all the position settings being passed around. The scroll strategy should be completely decoupled from that.

File: projects/igniteui-angular/src/lib/services/overlay/position/IPositionStrategy.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { PositionSettings } from './../utilities';
+
+export interface IPositionStrategy {
+    // _wrapperClass: string;
+    _settings: PositionSettings;
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/services/overlay/position/IPositionStrategy.ts]: This shouldn't exist on the interface. Assuming you do the target setting merge + util method each strategy can call in, the `_settings` becomes internal for each strategy as nothing else would need to use those, right?

File: projects/igniteui-angular/src/lib/services/overlay/utilities.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+export class Rectangle {
+    constructor(public x: number, public y: number, public w: number, public h: number) { }
+}
+
+export class Point {
+    constructor(public x: number, public y: number) { }
+}
+
+export class PositionSettings {
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/utilities.ts]: done

File: CHANGELOG.md [hunk 5]
```diff
@@ -41,7 +41,8 @@ export class IgxCustomFilteringOperand extends IgxFilteringOperand {
 
 - **Breaking changes**:
     - Removed submodule imports. All imports are now resolved from the top level `igniteui-angular` package.
-    - `igxToggle` changes
+    - `igxToggle` changes:
+        - `igxToggle` is renamed to `igxOverlay`.
```
Comment (kdinev) [inline CHANGELOG.md]: Wait, are you sure we need to rename the toggle? I don't quite remember when this decision was made!

File: projects/igniteui-angular/src/lib/services/overlay/position/auto-position-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        this.wrapperClass = 'auto-show';
+        const positioningStrat = new ConnectedPositioningStrategy(this._settings);
+        positioningStrat.position(element, wrapper, rect);
+        const viewPort = this.getViewPort(document);
+        const checkIfMoveHorizontal = (elem: HTMLElement) => {
+            const leftBound = elem.parentElement.offsetLeft;
+            const rightBound = elem.parentElement.offsetLeft + elem.clientWidth;
+            let newPosition;
+            switch (this._settings.horizontalDirection) {
+                case (-1):
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/position/auto-position-strategy.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/position/global-position-strategy.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+export class GlobalPositionStrategy implements IPositionStrategy {
+    public _settings: PositionSettings;
+
+    constructor(settings?: PositionSettings) {
+        this._settings = settings ? settings : new PositionSettings();
+    }
+
+    position(element: HTMLElement, wrapper: HTMLElement, size: {}): void {
+        switch (this._settings.horizontalDirection) {
+            case -1:
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/position/global-position-strategy.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/position/global-position-strategy.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            case 0:
+            wrapper.parentElement.style.justifyContent = 'flex-end';
+                break;
+            default:
+                break;
+        }
+
+        switch (this._settings.verticalDirection) {
+            case -1:
+            wrapper.parentElement.style.alignItems = 'flex-start';
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/position/global-position-strategy.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/utilities.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    constructor(
+        public point: Point = new Point(0, 0),
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/utilities.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/utilities.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        public horizontalDirection = HorizontalAlignment.Center,
+        public verticalDirection = VerticalAlignment.Middle,
+        public element?: HTMLElement,
+        public horizontalStartPoint: HorizontalAlignment = HorizontalAlignment.Center,
+        public verticalStartPoint: VerticalAlignment = VerticalAlignment.Middle) { }
+}
+
+export class OverlaySettings {
```
[the remark this answers is not in the document]
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/utilities.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/overlay.ts (hunk 1 above)
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/overlay.ts]: done

File: projects/igniteui-angular/src/lib/services/overlay/scroll/AbsoluteScrollStrategy.ts (hunk 2 above)
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/scroll/AbsoluteScrollStrategy.ts]: Reposition added to overlay.
We need reference to document, as we attach some events to it. As we cannot inject it through the constructor we are passing it via initialize method.

File: projects/igniteui-angular/src/lib/services/overlay/position/IPositionStrategy.ts (hunk 3 above)
Comment (wnvko) [reply projects/igniteui-angular/src/lib/services/overlay/position/IPositionStrategy.ts]: we need this at least to reach to animation in show and hide methods of overlay.
Also one can easily setup position settings via settings property, and not by providing settings via the constructor.

**Pull request merged**